image-prune

```diff
@@ -152,34 +140,74 @@ def prune_images(force, dryrun, quiet=False, open_pull_requests=True, offline=Fa
             target = os.readlink(entry.path)
             targets.add(target)
 
-    expiry_threshold = now - IMAGE_EXPIRE * 86400
-    for entry in os.scandir(get_images_data_dir()):
-        if not force and (enough_disk_space() and entry.stat(follow_symlinks=False).st_mtime > expiry_threshold):
-            continue
-        if entry.is_file() and \
-                any(entry.name.endswith(ext) for ext in ['.qcow2', '.partial']) and \
-                entry.name not in targets:
-            if not quiet or dryrun:
-                sys.stderr.write("Pruning {0}\n".format(entry.name))
+    return targets
+
+
+class ImageCache:
```

Comment:
This class should have abstract methods that are implemented by subclasses, such as `def delete_images()`, which raise a `NotImplementedError`. That will clarify and document what concrete subclasses need to implement.

Honestly, I find this whole OOification rather excessive here, but if you want to go through with it, let's do it properly.

Reply:
I read about this in depth last night and it seems that opinion is a bit split on the correct thing to do here.  If you have a preference, then I'm OK to do that.

As for OO: I was implementing the S3 stuff as a skeleton with this at the same time and it turned into a bunch of ugly cases of if S3: do_this() else: do_that() for the listing, the space determination and the deleting.  This screams for virtual methods.

Reply:
Right, as you have the non-OO comparison with the S3 part, I trust that this looks nicer.